src/from_fn.rs

```diff
@@ -2,66 +2,37 @@
 
 use crate::{Array, ArraySize};
 use core::{
+    convert::Infallible,
     mem::{self, MaybeUninit},
     ptr,
 };
 
-/// Construct an array type from the given generator function.
-pub trait FromFn<T>: Sized {
-    /// Create array using the given generator function for each element.
-    fn from_fn<F>(cb: F) -> Self
-    where
-        F: FnMut(usize) -> T;
-
-    /// Create an array using the given generator function for each element, returning any errors
-    /// which are encountered in the given generator.
-    ///
-    /// # Errors
-    ///
-    /// Propagates the `E` type from the provided `F` in the event of error.
-    fn try_from_fn<E, F>(f: F) -> Result<Self, E>
-    where
-        F: FnMut(usize) -> Result<T, E>;
-}
-
-impl<T, U> FromFn<T> for Array<T, U>
+impl<T, U> Array<T, U>
 where
     U: ArraySize,
 {
+    /// Create array where each array element `T` is returned by the `f` call.
     #[inline]
-    fn from_fn<F>(cb: F) -> Self
-    where
-        F: FnMut(usize) -> T,
-    {
-        Array::from_fn(cb)
-    }
-
-    #[inline]
-    fn try_from_fn<E, F>(cb: F) -> Result<Self, E>
-    where
-        F: FnMut(usize) -> Result<T, E>,
-    {
-        Array::try_from_fn(cb)
-    }
-}
-
-impl<T, const N: usize> FromFn<T> for [T; N] {
-    #[inline]
-    fn from_fn<F>(cb: F) -> Self
+    pub fn from_fn<F>(mut f: F) -> Self
     where
         F: FnMut(usize) -> T,
     {
-        core::array::from_fn(cb)
+        Self::try_from_fn::<Infallible, _>(|n| Ok(f(n))).expect("should never fail")
     }
 
-    // TODO(tarcieri): use `array::try_from_fn` when stable
-    fn try_from_fn<E, F>(cb: F) -> Result<Self, E>
+    /// Create array fallibly where each array element `T` is returned by the `f` call, or return
+    /// an error if any are encountered.
+    ///
+    /// # Errors
+    ///
+    /// Propagates the `E` type returned from the provided `F` in the event of error.
+    pub fn try_from_fn<E, F>(f: F) -> Result<Self, E>
     where
         F: FnMut(usize) -> Result<T, E>,
     {
         // SAFETY: an array of `MaybeUninit`s is always valid.
-        let mut array: [MaybeUninit<T>; N] = unsafe { MaybeUninit::uninit().assume_init() };
-        try_from_fn_erased(&mut array, cb)?;
+        let mut array: Array<MaybeUninit<T>, U> = unsafe { MaybeUninit::uninit().assume_init() };
+        try_from_fn_erased(array.0.as_mut(), f)?;
```

Comment:
Aah joy, this again:

```
error: this call for this type may be undefined behavior
  --> src/from_fn.rs:34:60
   |
34 |         let mut array: Array<MaybeUninit<T>, U> = unsafe { MaybeUninit::uninit().assume_init() };
   |                                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

It would probably be good to leverage [`Array::uninit`](https://docs.rs/hybrid-array/0.2.0-rc.3/hybrid_array/struct.Array.html#method.uninit) to construct this, but that would currently require adding the `ArraySize<ArrayType = [T; N]>` bound, which gets viral quickly and would impact several other APIs which call this, so I was trying to avoid it.

Reply:
Is it possible add this bound on the method itself?

Reply:
If the bound exists on the method then it needs to be added to everything which calls the method, e.g. the `FromIterator` impl.

I think I'd like to try a PR which finds a sane way to remove the bound from `Array::uninit` so we can use that, which also avoids triggering this clippy lint.

Reply:
I wasn't able to come up with a solution that satisfies the clippy lint, unfortunately.

https://github.com/RustCrypto/hybrid-array/pull/35/commits/7deb1a3d74a4506e2144d8d3c0bfc8a733a56214 removes the bounds on `ArrayType` and adds `allow(clippy::uninit_assumed_init)`.

It's not ideal, but allows `Array::try_from_fn` to use `Array::uninit` and `Array::assume_init`, so at least we only need to get things right in those two places.